### PR TITLE
Socket tests: don't retry CanceledByDispose tests on timeout

### LINKED ISSUE #42686: Socket Dispose during synchronous operation hangs on RedHat 7
Disposing a Socket that has a pending synchronous Connect/Accept operation intermittently hangs on RedHat7 and CentOS7 x64 in CI.

```c#
using var client = new Socket(SocketType.Stream, ProtocolType.Tcp);

Exception ex = await Assert.ThrowsAnyAsync<Exception>(async () =>
{
    Task connectTask = Task.Run(() => client.Connect(endPoint));
    await WaitForEventAsync(events, "ConnectStart"); // Wait until the operation starts
    Task disposeTask = Task.Run(() => client.Dispose());
    await new[] { connectTask, disposeTask }.WhenAllOrAnyFailed(); // Hang
});
```

Existing [Connect](https://github.com/dotnet/runtime/blob/1bc493ab8b65d167180ad1f41f3a6292d28d7094/src/libraries/System.Net.Sockets/tests/FunctionalTests/Connect.cs#L125) and [Accept](https://github.com/dotnet/runtime/blob/1bc493ab8b65d167180ad1f41f3a6292d28d7094/src/libraries/System.Net.Sockets/tests/FunctionalTests/Accept.cs#L293) tests are potentially masking this failure with the use of a timeout/retry combo.

Example [CI Console log](https://helixre107v0xdeko0k025g8.blob.core.windows.net/dotnet-runtime-refs-pull-42188-merge-8f4f5063b5cf4b0fba/System.Net.Sockets.Tests/console.de4ca856.log?sv=2019-02-02&se=2020-10-13T06%3A28%3A45Z&sr=c&sp=rl&sig=PxZsuxlyOrsTmWiLwX5U0Qjchb9GyLItifD4845u%2FIY%3D)

### PR DESCRIPTION
**Pull request by tmds**
These retries cause masking cases where the operation and
dispose Task are never completing.

Fixes https://github.com/dotnet/runtime/issues/42686#issuecomment-698767431

@MihaZupan @antonfirsov ptal

### CODE REVIEW

File: src/libraries/System.Net.Sockets/tests/FunctionalTests/Connect.cs [hunk 1]
```diff
@@ -127,7 +127,8 @@ public async Task ConnectGetsCanceledByDispose()
             // We try this a couple of times to deal with a timing race: if the Dispose happens
             // before the operation is started, we won't see a SocketException.
             int msDelay = 100;
-            await RetryHelper.ExecuteAsync(async () =>
+            int retries = 10;
+            while (true)
```
Comment (antonfirsov) [inline src/libraries/System.Net.Sockets/tests/FunctionalTests/Connect.cs:131]:
```C#
                catch (SocketException se)
                {
                    // On connection timeout, retry.
                    Assert.NotEqual(SocketError.TimedOut, se.SocketErrorCode);

                    localSocketError = se.SocketErrorCode;
                }
```

The test fails here on Windows. Should do some more research on `SocketError.TimedOut` cases, but I think we should keep retrying here.

**Review by antonfirsov — COMMENTED** [review]
Wouldn't it be better to extend `RetryHelper` to detect fatal failures?

See:
https://github.com/dotnet/runtime/commit/1339f52abb60c4af73306cd3776cb969ed898762

**Comment by antonfirsov** [comment]
> Fixes #42686 (comment)

I don't think that a possible test bug is the only concern in #42686.

**Comment by tmds** [comment]
> I don't think that a possible test bug is the only concern in #42686.

Yes, this is only meant to fix the test bug.

File: src/libraries/Native/Unix/System.Security.Cryptography.Native/pal_ssl.c [hunk 2]
```diff
@@ -118,6 +119,8 @@ static void DetectCiphersuiteConfiguration()
     g_config_specified_ciphersuites = 1;
 
 #endif
+    write(0, "config\n", 7);
+    write(0, g_config_specified_ciphersuites == 1 ? "1\n" : "0\n", 2);
```
Comment (antonfirsov) [inline src/libraries/Native/Unix/System.Security.Cryptography.Native/pal_ssl.c:123]: I guess you pushed this by accident?

Comment (tmds) [reply src/libraries/Native/Unix/System.Security.Cryptography.Native/pal_ssl.c]: yes, removed it.

**Comment by antonfirsov** [comment]
@tmds what do you think about my idea regarding `RetryHelper`? If we could simplify the changes by that, and fix the Windows failure, I'm happy to proceed merging this.

**Comment by tmds** [comment]
> @tmds what do you think about my idea regarding RetryHelper? If we could simplify the changes with that trick, and fix the Windows failure, I'm happy to proceed merging this.

I think you need to look at multiple tests to know whether it makes sense to extend the RetryHelper for this pattern.
I've moved the assert into the try-catch-retry section. It was the simplest way to fix the Windows failure.

**Comment by tmds** [comment]
Part of https://github.com/dotnet/runtime/pull/42725.

**Pull request closed without merging**